Python/Product/Analysis/Interpreter/Ast/AstAnalysisWalker.cs

```diff
@@ -0,0 +1,260 @@
+                return _interpreter.GetBuiltinType(BuiltinTypeId.Tuple);
+            }
+            if (expr is SetExpression || expr is SetComprehension) {
+                return _interpreter.GetBuiltinType(BuiltinTypeId.Set);
+            }
+            if (expr is LambdaExpression) {
+                return _interpreter.GetBuiltinType(BuiltinTypeId.Function);
+            }
+
+            return null;
+        }
+
+        private IMember GetInScope(string name) {
+            foreach (var m in _scope) {
+                if (m == null) {
+                    continue;
+                }
+                IMember obj;
+                if (m.TryGetValue("__class__", out obj)) {
+                    return obj;
+                }
+            }
+            return null;
+        }
+
+        private IPythonType CurrentClass {
+            get {
+                var m = _scope.Peek();
+                if (m != null) {
+                    IMember cls;
+                    m.TryGetValue("__class__", out cls);
+                    return cls as IPythonType;
+                }
+                return null;
+            }
+        }
+
+        public override bool Walk(AssignmentStatement node) {
+            var m = _scope.Peek();
+            if (m != null) {
+                var value = GetValueFromExpression(node.Right, m);
+                foreach (var ne in node.Left.OfType<NameExpression>()) {
+                    m[ne.Name] = value;
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(ImportStatement node) {
+            var m = _scope.Peek();
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i].Names[0];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(node.AsNames[i])
+                            );
+                        }
+                    }
+                } catch (IndexOutOfRangeException) {
+                }
+            }
+            return false;
+        }
+
+        public override bool Walk(FromImportStatement node) {
+            var m = _scope.Peek();
+            if (node.Root.MakeString() == "__future__") {
+                return false;
+            }
+
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(n)
+                            );
```

Comment:
Agreed. This was a lazy copy-paste from above. Ideally I'd like to be able to resolve these to class vs. function vs. constant from the referenced module (or at least submodules), since *many* libraries I looked at do relative imports for their public members, but it's not essential right now.

---
In reply to: [83938907](https://github.com/Microsoft/PTVS/pull/1729#discussion_r83938907) [](ancestors = 83938907)